File: Python/Medium/graphs/validateBinaryTree.py
import collections
def DFSSolution(n, leftChild, rightChild):
    graph = collections.defaultdict(list)
    has_parent = [False]*n

    for i in range(n):
        if leftChild[i] != -1:
            graph[i].append(leftChild[i])
            has_parent[leftChild[i]] = True

        if rightChild[i] != -1:
            graph[i].append(rightChild[i])
            has_parent[rightChild[i]] = True

    # Determine if there is a single root and what it is
    roots = [i for i in range(n) if not has_parent[i]]

    # If more than one root...not valid
    if len(roots) != 1:
        return False
    
    visited = [False]*n
    def DFS(node):
        if visited[node] == True:
            return False
        visited[node] = True

        for neighbor in graph[node]:
            if not DFS(neighbor):
                return False

        return True
    
    return DFS(roots[0]) and all(visited)


"""
DFS approach is slower and suboptimal for this problem.

BFS is better

"""

def BFSSolution(n, righChild, leftChild):
    in_degree = [0]*n

    for i in range(n):
        if leftChild[i] != -1:
            in_degree[leftChild[i]] += 1

        if righChild[i] != -1:
            in_degree[righChild[i]] += 1

    
    # find root
    root = None
    for i in range(n):
        if in_degree[i] == 0:
            if root is None:
                root = i
            else:
                return False
            
    if root is None:
        return False
    

    visited = [False]*n
    queue = [root]

    while queue:
        cur = queue.pop(0)
        if visited[cur] == True:
            return False
        visited[cur] = True
        if leftChild[cur] != -1:
            queue.append(leftChild[cur])

        if righChild[cur] != -1:
            queue.append(righChild[cur])

    return sum(visited) == n

File: JavaScript/Medium/graphs/validateBinaryTree.js
var DFSSolution = function(n, leftChild, rightChild){
    let graph = new Array(n).fill(0).map(() => [])
    let hasParent = new Array(n).fill(false)

    for (let i = 0; i<n; i++){
        if (leftChild[i] != -1){
            graph[i].push(leftChild[i])
            hasParent[leftChild[i]]=true
        }

        if (rightChild[i] != -1){
            graph[i].push(rightChild[i])
            hasParent[rightChild[i]]= true
        }
    }

    let roots = []
    for (let i = 0; i<n; i++){
        if (!hasParent[i]){
            roots.push(i)
        }
    }

    if (roots.length != 1) return false

    let visited = new Array(n).fill(false)
    var DFS = (node) => {
        if (visited[node]==true) return false
        visited[node] = true

        for (const neighbor of graph[node]){
            if (!DFS(neighbor)) return false
        }

        return true
    }

    return (DFS(roots[0]) && visited.every(v => v === true))
}



var BFSSolution = function(n, leftChild, rightChild){
    let inDegree = new Array(n).fill(0)

    for (let i = 0; i<n; i++){
        if (leftChild[i] != -1){
            inDegree[leftChild[i]] ++
        }

        if (rightChild[i] != -1){
            inDegree[rightChild[i]] ++
        }
    }


    let root = null;
    for (let i = 0; i<n; i++){
        if (inDegree[i] === 0){
            if (root === null){
                root = i
            }
            else {
                return false
            }
        }
    }

    if (root === null) return false

    let visited = new Array(n).fill(false)
    let queue = [root]

    while (queue.length){
        let cur = queue.shift()
        if (visited[cur] === true) return false

        visited[cur] = true

        if (leftChild[cur] != -1){
            queue.push(leftChild[cur])
        }

        if (rightChild[cur] != -1){
            queue.push(rightChild[cur])
        }
    }

    return visited.every(v => v === true)
}